python/ray/data/_internal/execution/operators/actor_pool_map_operator.py

```diff
@@ -290,8 +290,10 @@ def base_resource_usage(self) -> ExecutionResources:
         )
 
     def current_processor_usage(self) -> ExecutionResources:
-        # Both pending and running actors count towards our current resource usage.
-        num_active_workers = self._actor_pool.current_size()
+        # Exclude pending actors from the total count and only count running actors
+        # towards our current resource usage to avoid the progress bar from displaying
+        # more GPUs than actually used.
+        num_active_workers = self._actor_pool.num_running_actors()
```

Comment:
`current_processor_usage` is used by the Ray Data `ResourceManager` to determine resource budgets for operators. If we exclude pending tasks, I think the Ray Data scheduler might launch too many tasks

Reply:
good point. i think there are two ways we can work around this:

(1) In `StreamingExecutor._report_current_usage()`, instead of getting the CPU/GPU usage from `self._resource_manager.get_global_usage()`, we separately calculate the CPU/GPU usage using `self._actor_pool.num_running_actors()`. This leaves the global usage object untouched

(2) Modify the `ExecutionResources` object to store `active_cpu`, `pending_cpu`, `active_gpu`, `pending_gpu`, instead of lumping all the active + pending CPU/GPU usages into one attribute. This would be much more involved of a change.

@bveeramani do you have preference between the two options? I am thinking 1 is slightly better, because the refactor in 2 will be complicated, and not much other benefit that I can think of other than for this purpose.

Reply:
Thank you Balaji and Scott! I personally thought option 1 is a better approach as currently there isn't any other use case that will take advantage of the refactored `ExecutionResources` object. Let me know what your thought is, and I'll make adjustments to my code.

Reply:
We'd need to update the logic in two places: 
1. In `StreamingExecutor._report_current_usage` for the global usage. https://github.com/ray-project/ray/blob/8a7960fc70dddc5ae2c9a02352fba284f191be70/python/ray/data/_internal/execution/streaming_executor.py#L347
2. In `ResourceManager.get_op_usage_str` for the per-op usage. https://github.com/ray-project/ray/blob/8a7960fc70dddc5ae2c9a02352fba284f191be70/python/ray/data/_internal/execution/resource_manager.py#L183

>  instead of getting the CPU/GPU usage from self._resource_manager.get_global_usage(), we separately calculate the CPU/GPU usage using self._actor_pool.num_running_actors()

@scottjlee what would the code look like for this? Would it be something like this?

```
for op in ops:
    if isinstance(op, ActorPoolMapOperator):
        gpu_usage += op._actor_pool.num_runinng_actors()
    else:
        gpu_usage += op.current_processor_usage().gpu
```

I don't think it'd be ideal to have actor-specific logic in the executor code, but I also don't have any other ideas.

@raulchen could you weigh in on how we only display running actors in the progress bar? I'm stumped on how to set up the abstractions for this

Reply:
Agreed to avoid having actor-specific logic. 
One way to abstract this in my mind is to introduce a `pending resource usage` concept. Basically, we'll add a `pending_resource_usage()` method to `PhysicalOperator`, and the `ResourceManager` also keeps track of `global_pending_resource_usage`.

Reply:
@bveeramani yeah that's basically what i had in mind. I agree it's not ideal to have ActorPoolMapOperator specific logic here -- another approach is we could bubble up the `num_running_actors()` method to `PhysicalOperator` class, and return the appropriate value

Reply:
Thank you Balaji, Hao, and Scott for the feedback! I'm new to Ray, and I have learned a lot and gained more familiarity with the codebase with your guidance. I've adjusted my changes based on your feedback. Some redundancies are inevitable to avoid having `ActorPoolMapOperator` specific logic. If you have ideas on how to reduce the redundancies I introduced, please feel free to suggest!

Reply:
I think the updated approach looks good overall, but it doesn't explicitly define a `pending_resource_usage()` API. @raulchen was there another use case for this you had in mind? Or do you think the updated approach is good?

Reply:
The reason why I added an interface for running resources but not pending resources is that we want to exclude pending resources and only show running resources when displaying in use GPUs to the user. There isn't an explicit use case that will utilize pending resources directly.

Reply:
sorry for my late response.
I was thinking of updating the existing `current_processor_usage`  to `running_usage` and adding a new `pending_usage`. So we can should both running and pending in the progress bars. This seems more clear. 
What do you think?

And regarding other use cases (can be done in follow-up PRs):
* One to also move the [obj_store_mem_pending_task_outputs](https://github.com/ray-project/ray/blob/82af058f3456dd7088b94a8d29a95450f73d215b/python/ray/data/_internal/execution/interfaces/op_runtime_metrics.py#L338) method to the `pending_usage`.  ( discussed with @bveeramani last week)
* another is when an actor is preempted, we should update its usage to pending as well. (depending on a new API from core to implement this).

Reply:
@jeffreyjeffreywang following up on this, any thoughts on Hao's suggestion above? Feel free to ping me if you have any other questions on how to make the change, or want to discuss further (we can address the followups in separate PRs).

Reply:
Thanks for the suggestion, Hao! I didn't realize that we'd like to show pending usages as well in the progress bar. I agree that showing running and pending usages separately will be more clear rather than hiding pending usages from the progress bar. @scottjlee, I'll address the first suggestion in this PR as my current implementation may need to be modified. As for the follow-ups, I'll open a separate issue and address them in a follow up PR.

Reply:
Hey @scottjlee @raulchen, I've addressed your comments in my recent commit. Here's a screenshot of the progress bar after patching my changes:

![Screenshot 2024-08-13 at 4 40 33 PM](https://github.com/user-attachments/assets/870fd407-635b-4893-b5ee-01a00ed3077b)
![Screenshot 2024-08-13 at 4 21 13 PM](https://github.com/user-attachments/assets/282274e2-433d-4534-ba38-17c79859e569)

Note that the pending actors information is introduced on the first line of the summary along with running resources. The number of pending actors scheduled on GPUs starts from 1 and drops to 0 as expected. In `get_op_usage_str` where the resource usages for each operator is calculated and shown, I didn't add pending actors information to it in this PR as it is already included (Note that "1 actors (1 pending)" is displayed).

Please let me know if you think this output looks good to you. Again, thank you so much for your help along the way!

Reply:
The reason why I kept `current_processor_usage` is that it's used to calculate the global usage which is then used to determine whether an operator can be executed in `_execution_allowed` in `streaming_executor_state.py`. @raulchen, let me know if you think it's still better to get rid off `current_processor_usage`. Thank you 😄

Reply:
Opened issues #47125 #47126 to track enhancements proposed by @raulchen.

Reply:
Should we update `current_processor_usage()` to return `running_processor_usage() + pending_processor_usage()`? @raulchen WDYT?

Reply:
Yeah I was thinking of doing that.

Reply:
I implemented the logic for `current_processor_usage()` in `MapOperator` so that `ActorPoolMapOperator` and `TaskPoolMapOperator` don't have to implement the logic separately.

Reply:
Hmm it seems like updating `current_processor_usage()` to return `running_processor_usage() + pending_processor_usage()` will break `test_executor_resource_management` tests (specifically, `test_actor_pool_resource_reporting` and `test_actor_pool_resource_reporting_with_bundling`). Will revert it back to original behavior while I keep figuring out the reason.

Reply:
I prefer to let `running_processor_usage` return `current_processor_usage - pending_processor_usage` by default, for backward compatibility. 
if a test is broken, probably something is wrong. I can take a look at the failed tests.

Reply:
I tried the above. The test still passes. 
Let me know if other tests are failing.

Reply:
Thank you Hao. I was able to figure out the root cause of the failed tests when updating `current_processor_usage()` to return `running_processor_usage() + pending_processor_usage()`. I've updated the code to what you suggested regardless. Let me know if you prefer the `current_processor_usage() = running_processor_usage() + pending_processor_usage()` or `running_processor_usage() = current_processor_usage() - pending_processor_usage()`.

Reply:
can you keep `current_processor_usage` as it is and update `running_processor_usage` to return `current_processor_usage() - pending_processor_usage()`. 
This is better for backward compatibility. Because only a few ops need pending_processor_usage.

Reply:
No problem! It is now updated to what you have in mind.